crates/texlab/src/util.rs

```diff
@@ -11,13 +11,19 @@ pub fn normalize_uri(uri: &mut lsp_types::Url) {
         if let Some(mut path) = segments.next().and_then(fix_drive_letter) {
             for segment in segments {
                 path.push('/');
-                path.push_str(segment);
+                path.push_str(&segment);
             }
 
             uri.set_path(&path);
         }
     }
 
+    let encoded_path = uri.path();
+    let decoded_path = percent_encoding::percent_decode_str(encoded_path)
```

Comment:
I'm bit uncomfortable with this implementation. Two issues that come to mind:

- The `decode_utf8_lossy()` can't handle non-utf8 data. I'm not sure if that's supported otherwise (in LSP or URI spec, or elsewhere in the code), but I do know filesystem paths are not always utf-8.
- Decoding will lose any reserved characters (like `/`), which were encoded previously. For example `uri.set_path(decode("/a/b%2Fc"))` -> `"/a/b/c"`.


I was thinking if simple selective uppercasing would work:

- for each sequence of `%` followed by 2 hex characters
  - normalize the sequence to uppercase

This should keep exactly equivalent information, but normalize so that it's also byte by byte equal.

---

I do wonder, if the LSP clients mind the changing case? For example, if the client sends lower case encoded path to the server, and server later responds with uppercase encoded version. I don't think this _should_ matter, but I would not be surprised if it does for some client (like it did to this server).

Reply:
> I was thinking if simple selective uppercasing would work:

Agree, I changed the implementation to a simple loop where the next two characters after each `%` are converted to upper case.

> I do wonder, if the LSP clients mind the changing case?

I think, some of them do, although they _should_ not.

Reply:
Looks reasonable. The `percent-encoding` crate is no longer direct dependency for this implementation.

> I think, some of them do

I guess these would be better to fix in the client, than trying to accommodate those in the server.